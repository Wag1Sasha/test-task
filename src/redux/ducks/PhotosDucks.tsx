import { takeEvery, put, call } from 'redux-saga/effects';

export const REQUEST_PHOTOS_BY_VALUE = 'PHOTOS/REQUEST_PHOTOS_BY_VALUE';
export const ASYNC_FETCH_PHOTOS = 'PHOTOS/ASYNC_FETCH_PHOTOS';
export const ADD_PHOTO_TO_STORE = 'PHOTOS/ADD_PHOTO_TO_STORE';
export const SHOW_LOADER = 'PHOTOS/SHOW_LOADER';
export const HIDE_LOADER = 'PHOTOS/HIDE_LOADER';

const initialState = {
  fetchedPhotos: [],
  isLoading: false,
};

const photosReducer = (state = initialState, action: ReturnType<typeof addPhotoToStore>) => {
  switch (action.type) {
    case ADD_PHOTO_TO_STORE:
      return {
        ...state,
        fetchedPhotos: action.payload,
      };
    case SHOW_LOADER:
      return {
        ...state,
        isLoading: true,
      };
    case HIDE_LOADER:
      return {
        ...state,
        isLoading: false,
      };
    default:
      return state;
  }
};

export const addPhotoToStore = (photos: IPropsHits) => ({
  type: ADD_PHOTO_TO_STORE,
  payload: photos,
});

export const asyncFetchPhotos = (url: string) => ({
  type: ASYNC_FETCH_PHOTOS,
  url,
});

export const showLoader = () => ({
  type: SHOW_LOADER,
});

export const hideLoader = () => ({
  type: HIDE_LOADER,
});

export function* sagaWarcher() {
  yield takeEvery(ASYNC_FETCH_PHOTOS, sagaWorker);
}

function* sagaWorker(payload: ReturnType<typeof asyncFetchPhotos>) {
  const data = yield call(requestPhotos, payload.url);
  yield put(addPhotoToStore(data.hits));
}

const requestPhotos = async (url: string) => {
  const res = await fetch(`https://pixabay.com/api/?key=19776687-7c6cc257a41e8830f998f821c&q=${url}`);
  return await res.json();
};

export default photosReducer;
