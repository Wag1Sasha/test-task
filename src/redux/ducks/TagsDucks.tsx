export const EVENT_SUBMIT = 'TAGS/EVENT_SUBMIT';
export const DELETE_TAG = 'TAGS/DELETE_TAG';
export const DRAG_N_DROP = 'TAGS/DRAG_N_DROP';

const initialState: StateTagsType = {
  tags: [],
};

const tagsReducer = (state = initialState, action: TagsActionTypes) => {
  switch (action.type) {
    case EVENT_SUBMIT: {
      return {
        ...state,
        tags: [...state.tags, action.payload],
      };
    }
    case DELETE_TAG: {
      return {
        ...state,
        tags: [...state.tags.filter((_, i) => i !== action.payload)],
      };
    }
    case DRAG_N_DROP: {
      return {
        ...state,
        tags: [...action.payload],
      };
    }
    default:
      return state;
  }
};

export const eventSubmit = (tags: TagPropsType) => ({
  type: EVENT_SUBMIT,
  payload: tags,
});

export const dragNDrop = (tags: TagPropsType[]) => ({
  type: DRAG_N_DROP,
  payload: tags,
});

export const deleteTag = (index: number): TagsActionTypes => ({
  type: DELETE_TAG,
  payload: index,
});

export default tagsReducer;
