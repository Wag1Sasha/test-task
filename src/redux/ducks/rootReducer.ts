import tagsReducer from './TagsDucks';
import photosReducer from './PhotosDucks';
import { combineReducers } from 'redux';

const rootReducer = combineReducers({
  tags: tagsReducer,
  photos: photosReducer,
});

export default rootReducer;
