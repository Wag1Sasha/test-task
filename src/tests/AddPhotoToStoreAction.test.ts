import configureStore from 'redux-mock-store';
import * as selectActions from '../redux/ducks/PhotosDucks';

describe('Add Photo To Store Action', () => {
  const mockStore = configureStore();
  const store = mockStore();
  const photos = { largeImageURL: 'http://image.com/image.png' };

  beforeEach(() => {
    store.clearActions();
  });

  it('Action-creator works correct', () => {
    const expectedActions = [
      {
        type: 'PHOTOS/ADD_PHOTO_TO_STORE',
        payload: photos,
      },
    ];
    store.dispatch(selectActions.addPhotoToStore(photos));
    expect(store.getActions()).toEqual(expectedActions);
  });

  it('Render action-creator', () => {
    store.dispatch(selectActions.addPhotoToStore(photos));
    expect(store.getActions()).toMatchSnapshot();
  });
});
