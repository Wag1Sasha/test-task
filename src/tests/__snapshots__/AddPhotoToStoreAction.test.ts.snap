// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Photo To Store Action Render action-creator 1`] = `
Array [
  Object {
    "payload": Object {
      "largeImageURL": "http://image.com/image.png",
    },
    "type": "PHOTOS/ADD_PHOTO_TO_STORE",
  },
]
`;
